File: start-accessibility/python/workshop.py
from __future__ import annotations

from pathlib import Path
from urllib.parse import urlsplit

from copilot import define_tool
from copilot.rpc import PermissionDecisionApproveOnce, PermissionDecisionReject
from pydantic import BaseModel, Field

from accessibility_rule_catalog import ACCESSIBILITY_RULES

MAX_SNAPSHOT_BYTES = 1_000_000


class LookupParams(BaseModel):
    query: str = Field(description="The accessibility issue or WCAG criterion to look up.")


@define_tool(name="accessibility_rule_lookup", description="Looks up read-only WCAG guidance maintained by this application.", skip_permission=True)
def accessibility_rule_lookup(params: LookupParams) -> dict[str, object]:
    query = params.query.strip().lower()
    rule = next((item for item in ACCESSIBILITY_RULES if item.criterion.lower() in query or item.title.lower() in query or any(keyword in query for keyword in item.keywords)), None)
    if rule is None:
        return {"criterion": "No exact match", "title": "Criterion not found", "when_it_applies": "The issue is not represented in the workshop catalog.", "recommendation": "Verify the evidence and consult the complete WCAG reference."}
    return rule.__dict__


def create_snapshot_reader(working_directory: str):
    output_directory = Path(working_directory, ".playwright-mcp").resolve()
    existing = {path.resolve() for path in output_directory.glob("page-*.yml")} if output_directory.is_dir() else set()

    @define_tool(name="read_latest_accessibility_snapshot", description="Reads the newest Playwright accessibility snapshot created during this run.", skip_permission=True)
    def read_latest_accessibility_snapshot() -> str:
        candidates = [path for path in output_directory.glob("page-*.yml") if path.resolve() not in existing and not path.is_symlink() and path.is_file() and 0 < path.stat().st_size <= MAX_SNAPSHOT_BYTES]
        if not candidates:
            raise FileNotFoundError("No current-run Playwright snapshot is available. Call browser_navigate first.")
        return max(candidates, key=lambda path: path.stat().st_mtime).read_text(encoding="utf-8")

    return read_latest_accessibility_snapshot


def permission_for_target(target: str):
    def handler(request, _invocation):
        if getattr(request, "kind", None) == "mcp" and request.server_name == "playwright" and request.tool_name in {"browser_navigate", "playwright-browser_navigate"} and isinstance(request.args, dict) and isinstance(request.args.get("url"), str) and _same_url(request.args["url"], target):
            return PermissionDecisionApproveOnce()
        return PermissionDecisionReject(feedback="This workshop allows Playwright to navigate only to the exact requested target.")
    return handler


def _same_url(requested: str, allowed: str) -> bool:
    try:
        left, right = urlsplit(requested), urlsplit(allowed)
        return (left.scheme.lower(), left.hostname.lower() if left.hostname else "", left.port, left.username, left.password, left.path, left.query, left.fragment) == (right.scheme.lower(), right.hostname.lower() if right.hostname else "", right.port, right.username, right.password, right.path, right.query, right.fragment)
    except ValueError:
        return False


def report_prompt(target: str) -> str:
    return f"""次の URL を対象に、根拠に基づくアクセシビリティレビューを作成してください:  {target}.
1. `browser_navigate` を使って、指定された URL を開いてください。
2. `read_latest_accessibility_snapshot` を呼び出して、アクセシビリティツリーを確認してください。
3. スナップショットで確認できる、確度の高い課題を 3〜5 件特定してください。
4. 修正案を提案する前に、各課題について `accessibility_rule_lookup` を呼び出してください。

次の構成だけを返してください:
# Accessibility review
## Finding 1: <短い名称>
- Evidence: <ブラウザーで観測した具体的な要素またはページ構造>
- WCAG criterion: <カタログが返した基準とタイトル>
- Recommended remediation: <具体的な実装上の変更>
必要に応じて指摘事項のセクションを繰り返してください。
## Review limits
これはブラウザーで観測できる根拠に限定したレビューであり、WCAG 適合性の完全な監査ではないことを明記してください。
根拠を捏造したり、裏付けのない統計を報告したり、このページが WCAG に適合していると断定したりしないでください。"""

File: finished/nodejs/accessibility-report/src/workshop.ts
import { defineTool, type CopilotSession, type PermissionHandler } from "@github/copilot-sdk";
import { z } from "zod";
import { lstat, readdir, readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { accessibilityRules } from "./accessibility-rule-catalog.js";

const maxSnapshotBytes = 1_000_000;
const noMatch = { criterion: "No exact match", title: "Criterion not found", whenItApplies: "The issue is not represented in the workshop catalog.", recommendation: "Verify the evidence and consult the complete WCAG reference.", keywords: [] };

export const accessibilityRuleLookup = defineTool("accessibility_rule_lookup", {
  description: "Looks up read-only WCAG guidance maintained by this application.",
  parameters: z.object({ query: z.string().describe("The accessibility issue or WCAG criterion to look up.") }),
  skipPermission: true,
  handler: async ({ query }) => {
    const normalized = query.trim().toLowerCase();
    return accessibilityRules.find((rule) => normalized.includes(rule.criterion.toLowerCase()) || normalized.includes(rule.title.toLowerCase()) || rule.keywords.some((keyword) => normalized.includes(keyword))) ?? noMatch;
  },
});

export function createSnapshotReader(workingDirectory: string) {
  const outputDirectory = resolve(workingDirectory, ".playwright-mcp");
  const existingSnapshots = safeSnapshotNames(outputDirectory).then((names) => new Set(names.map((name) => resolve(outputDirectory, name))));
  return defineTool("read_latest_accessibility_snapshot", {
    description: "Reads the newest Playwright accessibility snapshot created during this run.",
    parameters: z.object({}),
    skipPermission: true,
    handler: async () => {
      const baseline = await existingSnapshots;
      const candidates = await Promise.all((await safeSnapshotNames(outputDirectory)).map(async (name) => {
        const path = resolve(outputDirectory, name);
        const details = await lstat(path);
        return { path, details };
      }));
      const snapshot = candidates
        .filter(({ path, details }) => !baseline.has(path) && !details.isSymbolicLink() && details.isFile() && details.size > 0 && details.size <= maxSnapshotBytes)
        .sort((left, right) => right.details.mtimeMs - left.details.mtimeMs)[0];
      if (!snapshot) throw new Error("No current-run Playwright snapshot is available. Call browser_navigate first.");
      return readFile(snapshot.path, "utf8");
    },
  });
}

async function safeSnapshotNames(directory: string): Promise<string[]> {
  try { return (await readdir(directory)).filter((name) => /^page-.*\.yml$/.test(name)); }
  catch { return []; }
}

export function permissionForTarget(target: URL): PermissionHandler {
  return (request) => {
    if (request.kind === "mcp" && request.serverName === "playwright" &&
      (request.toolName === "browser_navigate" || request.toolName === "playwright-browser_navigate")) {
      const args = request.args;
      const requestedUrl = args !== null && typeof args === "object" && !Array.isArray(args) &&
        typeof args.url === "string" ? args.url : undefined;
      if (requestedUrl && sameUrl(requestedUrl, target)) return { kind: "approve-once" };
    }
    return { kind: "reject", feedback: "This workshop allows Playwright to navigate only to the exact requested target." };
  };
}

function sameUrl(requested: string, allowed: URL): boolean {
  try {
    const parsed = new URL(requested);
    return parsed.protocol.toLowerCase() === allowed.protocol.toLowerCase() &&
      parsed.hostname.toLowerCase() === allowed.hostname.toLowerCase() &&
      parsed.port === allowed.port && parsed.username === allowed.username &&
      parsed.password === allowed.password && parsed.pathname === allowed.pathname &&
      parsed.search === allowed.search && parsed.hash === allowed.hash;
  } catch {
    return false;
  }
}

export async function streamResponse(session: CopilotSession, prompt: string): Promise<void> {
  await new Promise<void>((resolve, reject) => {
    let receivedDelta = false;
    const unsubscribe = session.on((event) => {
      if (event.type === "assistant.message_delta" && event.data.deltaContent) { receivedDelta = true; process.stdout.write(event.data.deltaContent); }
      else if (event.type === "assistant.message" && !receivedDelta) process.stdout.write(event.data.content);
      else if (event.type === "tool.execution_start") console.log(`\n[tool:start] ${event.data.toolName}`);
      else if (event.type === "tool.execution_complete") console.log(`[tool:done] success=${event.data.success}`);
      else if (event.type === "session.error") reject(new Error(event.data.message));
      else if (event.type === "session.idle") { console.log(); unsubscribe(); resolve(); }
    });
    void session.send({ prompt }).catch(reject);
  });
}

export function reportPrompt(target: URL): string {
  return `次の URL を対象に、根拠に基づくアクセシビリティレビューを作成してください:  ${target.href}.
1. browser_navigate を使って、指定された URL を開いてください。
2. read_latest_accessibility_snapshot を呼び出して、アクセシビリティツリーを確認してください。
3. スナップショットで確認できる、確度の高い課題を 3〜5 件特定してください。
4. 修正案を提案する前に、各課題について accessibility_rule_lookup を呼び出してください。

次の構成だけを返してください:
# Accessibility review
## Finding 1: <短い名称>
- Evidence: <ブラウザーで観測した具体的な要素またはページ構造>
- WCAG criterion: <カタログが返した基準とタイトル>
- Recommended remediation: <具体的な実装上の変更>
必要に応じて指摘事項のセクションを繰り返してください。
## Review limits
これはブラウザーで観測できる根拠に限定したレビューであり、WCAG 適合性の完全な監査ではないことを明記してください。
根拠を捏造したり、裏付けのない統計を報告したり、このページが WCAG に適合していると断定したりしないでください。`;
}
